universal-application-tool-0.0.1/app/controllers/applicant/RedirectController.java

```diff
@@ -0,0 +1,88 @@
+package controllers.applicant;
+
+import auth.ProfileUtils;
+import auth.UatProfile;
+import com.google.common.base.Preconditions;
+import com.google.common.base.Strings;
+import controllers.CiviFormController;
+import controllers.routes;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import javax.inject.Inject;
+import models.Applicant;
+import models.Program;
+import org.pac4j.play.java.Secure;
+import play.i18n.MessagesApi;
+import play.mvc.Http;
+import play.mvc.Result;
+import repository.ProgramRepository;
+import views.applicant.ApplicantUpsellCreateAccountView;
+
+public class RedirectController extends CiviFormController {
+  private final ProfileUtils profileUtils;
+  private final ProgramRepository programRepository;
+  private final ApplicantUpsellCreateAccountView upsellView;
+  private final MessagesApi messagesApi;
+
+  @Inject
+  public RedirectController(
+      ProfileUtils profileUtils,
+      ProgramRepository programRepository,
+      ApplicantUpsellCreateAccountView upsellView,
+      MessagesApi messagesApi) {
+    this.profileUtils = Preconditions.checkNotNull(profileUtils);
+    this.programRepository = Preconditions.checkNotNull(programRepository);
+    this.upsellView = Preconditions.checkNotNull(upsellView);
+    this.messagesApi = Preconditions.checkNotNull(messagesApi);
+  }
+
+  @Secure
+  public CompletableFuture<Result> programByName(Http.Request request, String programName) {
+    Optional<UatProfile> profile = profileUtils.currentUserProfile(request);
+    if (profile.isEmpty()) {
+      return CompletableFuture.completedFuture(
+          redirect(routes.CallbackController.callback("GuestClient")));
+    }
+    CompletableFuture<Applicant> applicant = profile.get().getApplicant();
+    CompletableFuture<Program> program = programRepository.getForSlug(programName);
+    return CompletableFuture.allOf(applicant, program)
+        .thenApply(
+            empty -> {
+              if (applicant.isCompletedExceptionally()) {
+                return notFound();
+              } else if (program.isCompletedExceptionally()) {
+                return notFound();
+              }
+              return redirect(
+                  controllers.applicant.routes.ApplicantProgramsController.edit(
+                      applicant.join().id, program.join().id));
+            });
+  }
+
+  @Secure
+  public CompletableFuture<Result> considerSignIn(Http.Request request, String redirectTo) {
```

Comment:
Suggestion (non-blocking): should this be considerRegister? or considerSignUp?

Reply:
Consider Register, yeah, that's better.